### DISCUSSION TITLE
How to use Ogmios in a private testnet?

### DISCUSSION THREAD

**Discussion by Et9797** [CardanoSolutions/ogmios#222] [Q & A] [answered]
I'm trying to set up Ogmios with a private testnet (https://github.com/woofpool/cardano-private-testnet-setup). I managed to make the container run and it seems to sync to the chain just fine. However, when I use the Typescript client connect to the Ogmios server and create an interaction context, it fails to connect and throws a socket hang up error
 `FetchError: request to http://0.0.0.0:1338/health failed, reason: socket hang up.` 
I've definitely set the port and host correctly, and it works normally when I connect to my other running instance of Ogmios that's syncing a mainnet node and is on another port (1337). 

```
const config = {
    connection: {
        host: '0.0.0.0',
        port: 1338,
        tls: false
    }
}

import { createInteractionContext,  } from '@cardano-ogmios/client';
import { InteractionContext } from '@cardano-ogmios/client';

const context: InteractionContext = await createInteractionContext(
    err => console.error(err),
    () => console.log("Connection closed."),
    { connection: config.connection }
)
console.log(context)

```
Synchronization seems to work fine
![image](https://user-images.githubusercontent.com/20364007/172433461-7b6f80c4-2fc7-4b42-8a05-a055a9df8d1c.png)
Error
![image](https://user-images.githubusercontent.com/20364007/172433286-ae1f54d2-8d86-4ccf-85c4-0ce0911233fb.png)

**Comment by KtorZ** [comment]
You shouldn't be using `0.0.0.0` for connecting a _client_. The semantic of that interface is a bit special (see: https://en.m.wikipedia.org/wiki/0.0.0.0) and only truly has a meaning when used on the server-side to listen on any address.

So, if your connecting to a container and correctly forwarding the container port to your host machine, I'd suggest to try connecting to "localhost" instead.

**Comment by Et9797** [comment]
> You shouldn't be using `0.0.0.0` for connecting a _client_. The semantic of that interface is a bit special (see: https://en.m.wikipedia.org/wiki/0.0.0.0) and only truly has a meaning when used on the server-side to listen on any address.
> 
> So, if your connecting to a container and correctly forwarding the container port to your host machine, I'd suggest to try connecting to "localhost" instead.

I've tried localhost, but still throws the same error though. I also noticed when starting the container message it displays dashboardUrl uses port 1337 for some reason? `{message: {"Server":{"dashboardUrl":"http://0.0.0.0:1337/", ...}`. 

What I am currently running:

```
docker run --rm \
   --name ogmios \
   -p 1338:1338 \
   -v ${PWD}/private-testnet:/testnet \
   cardanosolutions/ogmios:latest \
     --node-socket /testnet/node-bft1/node.sock \
     --node-config /testnet/configuration.yaml \
     --host 0.0.0.0
```

**Comment by Et9797** [comment]
Okay so I shut off my mainnet node + ogmios container and now set port 1337 (-p 1337:1337) and it works! Any idea what may be the culprit?

**Reply by rhyslbw** [reply, marked as answer]
> Any idea what may be the culprit?

The port mapping was invalid since Ogmios was being started on the default `1337`. The following would work:

```
docker run --rm \
   --name ogmios \
   -p 1338:1337 \
   -v ${PWD}/private-testnet:/testnet \
   cardanosolutions/ogmios:latest \
     --node-socket /testnet/node-bft1/node.sock \
     --node-config /testnet/configuration.yaml \
     --host 0.0.0.0
```

or

```
docker run --rm \
   --name ogmios \
   -p 1338:1338 \
   -v ${PWD}/private-testnet:/testnet \
   cardanosolutions/ogmios:latest \
     --node-socket /testnet/node-bft1/node.sock \
     --node-config /testnet/configuration.yaml \
     --host 0.0.0.0 \
     --port 1338
```

**Comment by AngelCastilloB** [comment]
Hi, I am having the same error using Ogmios with a private network, but changing the port number didn't resolve the issue in my case. is there any extra step you took @Et9797? thanks :)

**Reply by Et9797** [reply]
Hi Angel. Seems to still work for me, even when using the newest version v5.5.0 Ogmios docker container. 
![image](https://user-images.githubusercontent.com/20364007/177307088-21e65228-5e6d-400a-a21e-e9057ede80b1.png)

**Reply by AngelCastilloB** [reply]
Thanks @Et9797, it turns out I had an error on one of the genesis files, my startingTime was not in UTC, I changed it and its working fine now :)

**Comment by KtorZ** [comment]
As @rhyslbw pointed out, this was mostly an issue with the port mapping between your docker container and your host, not really related to Ogmios.

@Et9797 may you mark the discussion as answered, should everything's been answered now :)

**Reply by Et9797** [reply]
> @Et9797 may you mark the discussion as answered, should everything's been answered now :)
Does not appear to be an option to do that in this discussion
